library/psa_crypto_pake.c

```diff
@@ -0,0 +1,720 @@
+};
+
+enum psa_pake_sequence
+{
+    PSA_PAKE_SEQ_INVALID        = 0,
+    PSA_PAKE_X1_STEP_KEY_SHARE  = 1,    /* also X2S & X4S KEY_SHARE */
+    PSA_PAKE_X1_STEP_ZK_PUBLIC  = 2,    /* also X2S & X4S ZK_PUBLIC */
+    PSA_PAKE_X1_STEP_ZK_PROOF   = 3,    /* also X2S & X4S ZK_PROOF */
+    PSA_PAKE_X2_STEP_KEY_SHARE  = 4,
+    PSA_PAKE_X2_STEP_ZK_PUBLIC  = 5,
+    PSA_PAKE_X2_STEP_ZK_PROOF   = 6,
+    PSA_PAKE_SEQ_END            = 7,
+};
+
+#if defined(MBEDTLS_PSA_BUILTIN_PAKE)
+psa_status_t psa_pake_setup( psa_pake_operation_t *operation,
+                             const psa_pake_cipher_suite_t *cipher_suite)
+{
+    /* A context must be freshly initialized before it can be set up. */
+    if( operation->alg != 0 || operation->state != PSA_PAKE_STATE_INVALID )
+    {
+        return( PSA_ERROR_BAD_STATE );
+    }
+
+    if( cipher_suite == NULL ||
+        PSA_ALG_IS_PAKE(cipher_suite->algorithm ) == 0 ||
+        ( cipher_suite->type != PSA_PAKE_PRIMITIVE_TYPE_ECC &&
+          cipher_suite->type != PSA_PAKE_PRIMITIVE_TYPE_DH ) ||
+        PSA_ALG_IS_HASH( cipher_suite->hash ) == 0 )
+    {
+        return( PSA_ERROR_INVALID_ARGUMENT );
+    }
+
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_ECJPAKE)
+    if( cipher_suite->algorithm != PSA_ALG_JPAKE ||
+        cipher_suite->type != PSA_PAKE_PRIMITIVE_TYPE_ECC ||
+        cipher_suite->family != PSA_ECC_FAMILY_SECP_R1 ||
+        cipher_suite->bits != 256 ||
+        cipher_suite->hash != PSA_ALG_SHA_256 )
+    {
+        return( PSA_ERROR_NOT_SUPPORTED );
+    }
+
+    operation->alg = cipher_suite->algorithm;
+
+    mbedtls_ecjpake_init( &operation->ctx.ecjpake );
+
+    operation->state = PSA_PAKE_STATE_SETUP;
+    operation->sequence = PSA_PAKE_SEQ_INVALID;
+    operation->input_step = PSA_PAKE_STEP_X1_X2;
+    operation->output_step = PSA_PAKE_STEP_X1_X2;
+
+    operation->buffer = NULL;
+    operation->buffer_length = 0;
+    operation->buffer_offset = 0;
+
+    return( PSA_SUCCESS );
+#else
+    return( PSA_ERROR_NOT_SUPPORTED );
+#endif
+}
+
+psa_status_t psa_pake_set_password_key( psa_pake_operation_t *operation,
+                                        mbedtls_svc_key_id_t password )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_attributes_t attributes = psa_key_attributes_init();
+    psa_key_type_t type;
+    psa_key_usage_t usage;
+
+    if( operation->alg == 0 ||
+        operation->state != PSA_PAKE_STATE_SETUP )
+    {
+        return( PSA_ERROR_BAD_STATE );
+    }
+
+    status = psa_get_key_attributes( password, &attributes );
+    if( status != PSA_SUCCESS )
+        return status;
+
+    type = psa_get_key_type( &attributes );
+    usage = psa_get_key_usage_flags( &attributes );
+
+    psa_reset_key_attributes( &attributes );
+
+    if( type != PSA_KEY_TYPE_PASSWORD &&
+        type != PSA_KEY_TYPE_PASSWORD_HASH )
+    {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+
+    if( usage == 0 ||
+        ( usage & PSA_KEY_USAGE_DERIVE ) == 0 )
```

Comment:
I hope I'm not saying something stupid, but it seems to me that the first condition is redundant, surely if `usage` is 0 then the result of `&`ing it with something will be 0 too?

Reply:
Indeed, it must be a remain of the previous reworks